File: LC2487_Remove_Nodes_from_Linked_List/remove_nodes_from_linked_list.cpp
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

#include <iostream>
using namespace std;

class Solution {
public:
    ListNode* rRemoveNodes(ListNode* curr)
    {
        if (curr->next == nullptr)
        {
            // base case
            return curr;
        }
        curr->next = this->rRemoveNodes(curr->next);
        if (curr->val < curr->next->val)
        {
            // if we need to remove curr
            return curr->next;
        }
        else
        {
            // no need to remove curr
            return curr;
        }
    }

    ListNode* removeNodes(ListNode* head) {
        return this->rRemoveNodes(head);
    }
};

File: LC2487_Remove_Nodes_from_Linked_List/remove_nodes_from_linked_list.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def __init__(self):
        self.max_right = 0; # largest value we see to the right

    def rRemoveNodes(self, curr_node: ListNode, prev: ListNode) -> ListNode:
        if not curr_node.next: # base case
            self.max_right = curr_node.val
            return curr_node
        else:
            self.rRemoveNodes(curr_node.next, curr_node)
            if curr_node.val < self.max_right:
                # Remove curr_node if necessary.
                if not prev:
                    # if we need to delete curr_node but it's the first node of the list
                    return curr_node.next
                else:
                    prev.next = curr_node.next
                    return prev.next
            else:
                # no need to remove curr_node
                self.max_right = curr_node.val
                return curr_node

    def removeNodes(self, head: Optional[ListNode]) -> Optional[ListNode]:
        return self.rRemoveNodes(head, None)